bpf/lib/nodeport.h

```diff
@@ -1176,55 +1177,56 @@ static __always_inline bool snat_v4_needed(struct __ctx_buff *ctx, __be32 *addr,
 #endif
 
 	ep = __lookup_ip4_endpoint(ip4->saddr);
-	if (ep && !(ep->flags & ENDPOINT_F_HOST)) {
-		struct remote_endpoint_info *info;
-		*from_endpoint = true;
+	/* If this is not a local endpoint, or if this is a localhost endpoint,
+	 * no SNAT is needed.
+	 */
+	if (!ep || (ep->flags & ENDPOINT_F_HOST))
+		return false;
```

Comment:
Thinking out loud: this branch is very unlikely to happen, as for majority of cases it's going to be from pod or from host traffic. Also, the check against `IPV4_MASQUERADE` above should catch the localhost endpoint case for majority of cases. Maybe it's a good use case for `unlikely()`?

Reply:
I can add it. I've seen it being used elsewhere in the datapath, but have no idea if it generally helps or not.

Reply:
I've seen noticeable improvements on DPDK code. Never checked on BPF. Compiler definitely takes it into account since we've since it fix a complexity issue once (just because code was reordered).

Reply:
I would leave this case probably as is rather than marking as unlikely() given afaiu this could be hit under multi-dev case.